### PR TITLE
schedule: actively push operator 

### PR DESCRIPTION
**Pull request by nolouch**
### What problem does this PR solve? 
For https://github.com/tikv/tikv/pull/4591 and speed up scheduling, we need actively push the operator.

### What is changed and how it works?
use a heap to get the needed push operator.

### Check List 

Tests 

 - Unit test
 - Manualy test
   TiKV use this branch: https://github.com/tikv/tikv/pull/4684
   Without the improve: 
   Screen Shot 2019-05-16 at 12 51 58 AM
   The branch:
    Screen Shot 2019-05-16 at 12 53 48 AM

### CODE REVIEW

File: server/schedule/operator_controller.go [hunk 1]
```diff
@@ -83,6 +92,61 @@ func (oc *OperatorController) Dispatch(region *core.RegionInfo) {
...
+	oc.Lock()
+	defer oc.Unlock()
+	if oc.opNotifierQueue.Len() == 0 {
+		return nil, false
+	}
+	item := oc.opNotifierQueue[0]
+	regionID := item.op.regionID
+	op, ok := oc.operators[regionID]
+	if !ok || op == nil {
+		heap.Pop(&oc.opNotifierQueue)
```
Comment (disksing) [inline server/schedule/operator_controller.go]: I think you can first pop it, then push it back to the heap if it is still pending.

File: server/schedule/operator_queue.go [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package schedule
+
+type operatorWithTime struct {
+	op    *Operator
+	time  int64
+	index int
```
Comment (disksing) [inline server/schedule/operator_queue.go]: Why need `index`?

File: server/schedule/operator_queue.go [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package schedule
+
+type operatorWithTime struct {
+	op    *Operator
+	time  int64
```
Comment (disksing) [inline server/schedule/operator_queue.go]: Why not use `time.Time` directly?

File: server/schedule/operator_controller.go [hunk 4]
```diff
@@ -83,6 +99,61 @@ func (oc *OperatorController) Dispatch(region *core.RegionInfo) {
...
+	item.time = oc.getNextPushOperatorTime(step, now)
+	heap.Push(&oc.opNotifierQueue, item)
+	return r, true
+}
+
+// PushOperators periodically pushs the unfinished operator to the excutor(TiKV).
+func (oc *OperatorController) PushOperators() {
+	for {
+		r, next := oc.pollNeedDispatchRegion()
+		if r == nil && !next {
```
Comment (rleungx) [inline server/schedule/operator_controller.go]: How about:
```
if r == nil {
  	if !next {
		break
	}
	continue
}
```

File: server/schedule/operator_controller.go [hunk 5]
```diff
@@ -83,6 +99,61 @@ func (oc *OperatorController) Dispatch(region *core.RegionInfo) {
...
+		return nil, false
+	}
+
+	// pushs with new notify time.
+	item.time = oc.getNextPushOperatorTime(step, now)
+	heap.Push(&oc.opNotifierQueue, item)
+	return r, true
+}
+
+// PushOperators periodically pushs the unfinished operator to the excutor(TiKV).
```
Comment (rleungx) [inline server/schedule/operator_controller.go]:
```suggestion
// PushOperators periodically pushes the unfinished operator to the executor(TiKV).
```

File: server/schedule/operator_queue.go [hunk 6]
```diff
@@ -0,0 +1,46 @@
+// Copyright 2018 PingCAP, Inc.
```
Comment (rleungx) [inline server/schedule/operator_queue.go]: 2019

File: server/schedule/operator_controller.go [hunk 7]
```diff
@@ -83,6 +99,61 @@ func (oc *OperatorController) Dispatch(region *core.RegionInfo) {
...
+	if step == nil {
+		return nil, true
+	}
+	now := time.Now()
+	if now.Before(item.time) {
+		heap.Push(&oc.opNotifierQueue, item)
+		return nil, false
+	}
+
+	// pushs with new notify time.
```
Comment (rleungx) [inline server/schedule/operator_controller.go]: pushes

File: server/schedule/operator_controller_test.go [hunk 8]
```diff
@@ -84,11 +85,58 @@ func (t *testOperatorControllerSuite) TestOperatorStatus(c *C) {
...
+	tc.AddLeaderRegion(2, 1, 2)
+	steps := []OperatorStep{
+		RemovePeer{FromStore: 2},
+		AddPeer{ToStore: 2, PeerID: 4},
+	}
+	op1 := NewOperator("test", 1, &metapb.RegionEpoch{}, OpRegion, TransferLeader{ToStore: 2})
+	op2 := NewOperator("test", 2, &metapb.RegionEpoch{}, OpRegion, steps...)
+	region1 := tc.GetRegion(1)
+	region2 := tc.GetRegion(2)
+	// Adds operator and pushs to the notifier queue.
```
Comment (rleungx) [inline server/schedule/operator_controller_test.go]: pushes

File: server/schedule/operator_controller.go [hunk 9]
```diff
@@ -83,6 +99,62 @@ func (oc *OperatorController) Dispatch(region *core.RegionInfo) {
...
+	item.time = oc.getNextPushOperatorTime(step, now)
+	heap.Push(&oc.opNotifierQueue, item)
+	return r, true
+}
+
+// PushOperators periodically pushes the unfinished operator to the executor(TiKV).
+func (oc *OperatorController) PushOperators() {
+	for {
+		r, next := oc.pollNeedDispatchRegion()
+		if r == nil {
```
Comment (shafreeck) [inline server/schedule/operator_controller.go]: Personal suggestion
```go
if !next {
    break
}
if r == nil {
    continue
}
```

`next` is an independent return value and it does not depend on the value of `r`, `next` controls the loop routine whatever the value of `r` is.

Maybe this is more acceptable when you think this as something like this: `val, err := foo()`

File: server/schedule/operator_queue.go [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+
+func (opn *operatorQueue) Push(x interface{}) {
+	item := x.(*operatorWithTime)
+	*opn = append(*opn, item)
+}
+
+func (opn *operatorQueue) Pop() interface{} {
+	old := *opn
+	n := len(old)
+	item := old[n-1]
```
Comment (shafreeck) [inline server/schedule/operator_queue.go]: This causes a panic when n is 0

https://play.golang.org/p/qV3sTknOkb1

File: server/coordinator.go [hunk 11]
```diff
@@ -131,6 +131,23 @@ func (c *coordinator) patrolRegions() {
...
+	defer logutil.LogPanic()
+
+	defer c.wg.Done()
+	log.Info("coordinator begins to actively drive push operator")
+	for {
+		select {
+		case <-c.ctx.Done():
+			log.Info("drive push operator has been stopped")
+			return
+		case <-time.After(schedule.PushOperatorTickInterval):
```
[the remark this answers is not in the document]
Comment (nolouch) [reply server/coordinator.go]: got

[the remark this answers is not in the document]
Comment (shafreeck) [reply server/coordinator.go]: @nolouch  You are right, it will be collected after being fired.

**Pull request merged**